web/src/desktop_notifications.ts

```diff
@@ -0,0 +1,99 @@
+import $ from "jquery";
+
+import type {Message} from "./message_store";
+
+export type NoticeMemEntry = {
+    obj: ElectronBridgeNotification | Notification;
+    msg_count: number;
+    message_id: number;
+};
+
+export const notice_memory: Map<string, NoticeMemEntry> = new Map();
+
+export let NotificationAPI: typeof ElectronBridgeNotification | typeof window.Notification;
+
+export function set_notification_api(
+    n: typeof ElectronBridgeNotification | typeof window.Notification,
+): void {
+    NotificationAPI = n;
+}
+
+class ElectronBridgeNotification extends EventTarget {
+    constructor(title: string, options: NotificationOptions) {
+        super();
+        if (window.electron_bridge && window.electron_bridge.new_notification) {
+            Object.assign(
+                this,
+                window.electron_bridge.new_notification(title, options, (type, eventInit) =>
+                    this.dispatchEvent(new Event(type, eventInit)),
+                ),
+            );
+        }
+    }
+
+    static get permission(): NotificationPermission {
+        return Notification.permission;
+    }
+
+    static async requestPermission(
+        callback: NotificationPermissionCallback | undefined = undefined,
+    ): Promise<NotificationPermission> {
+        if (callback) {
+            callback(await Promise.resolve(Notification.permission));
+        }
+        return Notification.permission;
+    }
+
+    close(): void {
+        if (Object.prototype.hasOwnProperty.call(this, "close")) {
+            this.close();
+        }
+    }
```

Comment:
This added wrapper code still doesn’t make any sense. If it’s being invoked, it’s because `.close` has already resolved to this wrapper, so there are no circumstances under which calling `this.close()` again will do anything other than re-invoke the same wrapper.

Since you have added it to appease the TypeScript checker, the right course of action is to go back and ask yourself why there was a problem that needed to be appeased. The answer is that the `Object.assign` call in the constructor is completely unchecked; TypeScript cannot infer any type information from it. So the solution is to replace that `Object.assign` call with individual assignments that TypeScript understands.

Reply:
Thanks for providing the solution!

I will replace ```Object.assign()``` with individual assignment. To assign individually, I first need to define all keys explicitly that the class will consist. It will not be hard because keys will same as the keys of type ```NotificationData```.

Reply:
I just realise that if we only define all the keys explicitly, only that will solve the problem. We don't have to do individual assignment. ```Object.assign()``` will work.

Reply:
`Object.assign()` is still unchecked, and we don’t want to have unchecked code. And if you weren’t filling the class with phony defaults to appease TypeScript, it would be complaining anyway.